### ISSUE TITLE
Handle Mixed message templates better, e.g. {0} {Order}

### ISSUE THREAD

**Issue by 304NotModified** [NLog/NLog#2532]
I'm working on the docs and examples for NLog 4.5

Maybe we discussed this before, but I'm trying to use it as a regular NLog user.

**Type** (choose one):

- Bug / feature

**NLog version**:  4.5 rc4

**Platform**:  all platforms, tested on .NET Core

**Current NLog config** (xml or C#, if relevant)

```c#

   var order = new Order
            {
                ClientId = 301,
                OrderId = 2500002,
                Remarks = "My First order"
            };

logger.Info("Process order {0} for {@Order}", "CoolClient", order);
```

In case of a BUG:

- What is the current result?

No formatting as all of the message, no properties

The message is currently ` Process order {0} for {@Order}`

- What is the expected result?

Handle as structured template and create properties.

expected result:

```
 Process order "CoolClient" for {"OrderId":2500002, "ClientId":301, "Remarks":"My First order", "Express":false, "Status":"Processing"}
```

NB: `    Process order {@0} for {@Order}` also won't work

- Are there any work arrounds? no

**Comment by 304NotModified** [comment]
@snakefoot please your opinion. I could add an unit test if you agree on this.

**Comment by snakefoot** [comment]
Mixed mode should work when having parseMessageTemplate=true

Please create unit trst and fix code if not


Sent from my Samsung device


-------- Original message --------

**Comment by 304NotModified** [comment]
and with parseMessageTemplate=auto?

**Comment by snakefoot** [comment]
Then you want performance and none breaking behaviour



Sent from my Samsung device


-------- Original message --------

**Comment by 304NotModified** [comment]
true,

as `Process order {0} for {Order}` wasn't allowed first, I don't think would be a breaking change

**Comment by 304NotModified** [comment]
The rules are clear, really like to hold on that:



> - Templates that use numeric property names like {0} and {1} exclusively imply that arguments to the template are captured by numeric index
> - If any of the property names are non-numeric, then all arguments are captured by matching left-to-right with holes in the order in which they appear

from https://messagetemplates.org/

Any idea how to fix bullet 2, if the whole parse is now lazy? Could we do a non-lazy parse if we detected is a non-string.format?

**Comment by snakefoot** [comment]
>  Any idea how to fix bullet 2, if the whole parse is now lazy?

Should not be impossible to do an optimistic parse, and then do a full-parse if detecting a mixed mode situation. All the logic you implemented is still there, have not touched it.

**Comment by 304NotModified** [comment]
> All the logic you implemented is still there, have not touched it.

Yes, I saw #2541 was still there ;)



> Should not be impossible to do an optimistic lazy parse, and then do a full-parse if detecting a mixed mode situation.

True, not impossible, but don't wanna undo the performance optimizations

**Comment by 304NotModified** [comment]
another option would be to handle things differently as soon as we detect the "mixed mode".

But then I need to "undo" or postpone the stringbuilder write. 

What do you think?

**Comment by snakefoot** [comment]
> True, not impossible, but don't wanna undo the performance optimizations

Well the performance optimization is to lazy parse without allocating a list of MessageTemplateParameters, when normal numeric-string-format-string. Instead it just render numeric-parameters directly.

Just have a counter of how many numeric parameters that has been parsed, and use that offset for the named-parameter-index.

**Comment by 304NotModified** [comment]
> Well the performance optimization is to lazy parse without allocating a list of MessageTemplateParameters

And what about allocating a list of holes? If that also unwanted?

**Comment by snakefoot** [comment]
> And what about allocating a list of holes? Is that also unwanted?

The list should only contain named parameters, not the numeric ones.

**Comment by snakefoot** [comment]
You can also just change parseMessageTemplates=true, so it always parses as message-templates, so {0}, {1} becomes full properties. But then you will get into the funny business that Serilog fights with trying to ignore them afterwards.

Can see that Microsoft Extension Logging doesn't recognize numeric-templates, but just treats all as structured-templates (But it hides the parameter names, but users might be surprised that the numeric numbers has no effect).

**Comment by 304NotModified** [comment]
fixed by #2544
